File: backend/services/gitArchaeology.js
import simpleGit from 'simple-git';
import { getChatModel } from './llmProvider.js';

const LAZY_COMMIT_PATTERNS = [
  /^fix$/i, /^wip$/i, /^update$/i, /^updates$/i, /^asdf$/i,
  /^test$/i, /^temp$/i, /^changes$/i, /^bugfix$/i, /^done$/i,
  /^minor$/i, /^commit$/i, /^clean$/i, /^refactor$/i, /^misc$/i
];

/**
 * Extract Git commit history and synthesize AI Intent Summaries from raw diffs
 * Batches lazy commits into a SINGLE LLM prompt to prevent rate limits!
 */
export async function analyzeGitArchaeology(repoPath, maxCommits = 15) {
  const git = simpleGit(repoPath);

  let isRepo = false;
  try {
    isRepo = await git.checkIsRepo();
  } catch (e) {
    isRepo = false;
  }

  if (!isRepo) {
    console.log(`[Git Archaeology] "${repoPath}" is not a Git repository. Skipping commit history.`);
    return [];
  }

  let log = null;
  try {
    log = await git.log({ maxCount: maxCommits });
  } catch (err) {
    console.warn('[Git Archaeology] Could not read git log:', err.message);
    return [];
  }

  const results = [];
  const lazyCommits = [];

  for (const commit of log.all) {
    let diff = '';
    try {
      diff = await git.show([commit.hash, '--stat', '-p']);
      diff = diff.slice(0, 1500);
    } catch (e) {
      diff = '';
    }

    const isLazy = isLazyMessage(commit.message);
    const shortHash = commit.hash.slice(0, 7);

    const record = {
      hash: shortHash,
      fullHash: commit.hash,
      author: commit.author_name,
      date: commit.date,
      rawMessage: commit.message,
      intentSummary: commit.message,
      isSynthesized: isLazy,
      diffSnippet: diff.slice(0, 800)
    };

    results.push(record);

    if (isLazy && diff.length > 50) {
      lazyCommits.push({
        hash: shortHash,
        rawMessage: commit.message,
        diffSnippet: diff.slice(0, 1200)
      });
    }
  }

  // Single batched LLM call for all lazy commits to conserve rate limits
  if (lazyCommits.length > 0) {
    try {
      const chatModel = await getChatModel();
      if (chatModel) {
        const intentMap = await batchSynthesizeDiffIntents(chatModel, lazyCommits);
        for (const item of results) {
          if (intentMap[item.hash]) {
            item.intentSummary = intentMap[item.hash];
          }
        }
      }
    } catch (e) {
      console.warn('[Git Archaeology] Batch intent synthesis skipped:', e.message);
    }
  }

  return results;
}

/**
 * Checks if a commit message is uninformative
 */
function isLazyMessage(msg) {
  const trimmed = (msg || '').trim();
  if (trimmed.length <= 8) return true;
  for (const pattern of LAZY_COMMIT_PATTERNS) {
    if (pattern.test(trimmed)) return true;
  }
  return false;
}

/**
 * Single batched call to synthesize intent for all lazy commits at once
 */
async function batchSynthesizeDiffIntents(chatModel, lazyCommits) {
  const commitsPrompt = lazyCommits.map(c => 
    `Commit [${c.hash}] (Message: "${c.rawMessage}"):\nDiff snippet:\n${c.diffSnippet}`
  ).join('\n---\n');

  const prompt = `You are a Principal Software Architect performing Git archaeology.
The following commits have vague or uninformative commit messages.
For each commit, analyze its diff snippet and provide a 1-sentence summary of the actual architectural intent and what changed.

${commitsPrompt}

Respond strictly with valid JSON mapping each commit hash to its 1-sentence architectural summary, example:
{
  "${lazyCommits[0].hash}": "Architectural explanation of this change."
}`;

  try {
    const response = await chatModel.invoke(prompt);
    const text = typeof response.content === 'string' ? response.content : response.content?.[0]?.text || '';
    const cleanJson = text.replace(/```json/gi, '').replace(/```/g, '').trim();
    return JSON.parse(cleanJson);
  } catch (err) {
    return {};
  }
}

File: backend/services/git_archaeology.py
import os
import logging
from typing import List, Dict, Any
from git import Repo
from .llm_provider import get_chat_model

logger = logging.getLogger("git_archaeology")

async def analyze_git_archaeology(repo_path: str, max_commits: int = 15) -> List[Dict[str, Any]]:
    diff_summaries = []
    try:
        repo = Repo(repo_path)
        if not repo.heads:
            return []
        
        commits = list(repo.iter_commits(max_count=max_commits))
    except Exception as e:
        logger.warning(f"Git archaeology unavailable for {repo_path}: {e}")
        return []

    # A forced re-index must not wait for up to one cloud LLM request per
    # commit. Commit messages and diff snippets are already useful grounded
    # archaeology context. Opt into costly generated summaries explicitly.
    chat_model = None
    if os.getenv("GIT_ARCHAEOLOGY_LLM_SUMMARIES", "false").lower() in {"1", "true", "yes"}:
        try:
            chat_model = get_chat_model(temperature=0.1)
        except Exception:
            pass

    for commit in commits:
        try:
            h = commit.hexsha[:8]
            msg = commit.message.strip()
            author = commit.author.name or "Unknown"
            date = str(commit.committed_datetime)

            # Get diff with parent
            diff_text = ""
            if commit.parents:
                parent = commit.parents[0]
                diffs = parent.diff(commit, create_patch=True)
                patches = []
                for d in diffs[:5]:
                    if d.diff:
                        try:
                            patches.append(d.diff.decode('utf-8', errors='ignore'))
                        except Exception:
                            pass
                diff_text = "\n".join(patches)[:2000]
            else:
                diff_text = "Initial commit."

            intent = f"Commit {h}: {msg}"
            is_synthesized = False

            # Synthesize intent using LangChain
            if chat_model and diff_text:
                try:
                    prompt = (
                        f"Summarize the architectural and technical intent of this Git commit in 2-3 sentences:\n"
                        f"Commit Message: {msg}\n"
                        f"Diff Snippet:\n{diff_text[:1500]}\n\n"
                        f"Intent Summary:"
                    )
                    resp = await chat_model.ainvoke(prompt)
                    summary = resp.content if isinstance(resp.content, str) else str(resp.content)
                    if summary:
                        intent = summary.strip()
                        is_synthesized = True
                except Exception as llm_err:
                    logger.debug(f"LLM commit intent synthesis skipped: {llm_err}")

            diff_summaries.append({
                "hash": h,
                "author": author,
                "date": date,
                "rawMessage": msg,
                "diffSnippet": diff_text[:1000],
                "intentSummary": intent,
                "isSynthesized": is_synthesized
            })
        except Exception as err:
            logger.warning(f"Error processing commit {commit.hexsha}: {err}")

    return diff_summaries
